svscan.8
.TH svscan 8
.SH NAME
svscan \- starts and monitors a collection of services
.SH SYNOPSIS
.B svscan
[
.I directory
] [
.I log-service
]
.SH DESCRIPTION
.B svscan
starts one
.BR supervise (8)
process for each subdirectory of the current directory, up to a limit of 1000
subdirectories.
.B svscan
skips subdirectory names starting with dots.
.BR supervise (8)
must be in
.BR svscan 's
path.

.B svscan
optionally starts a pair of
.BR supervise (8)
processes, one for a subdirectory
.IR s ,
one for
.IR s\fR/\fBlog ,
with a pipe between them. It does this if the name
.I s
is at most 255 bytes long and
.I s\fR/\fBlog
exists and is a directory. (In versions 0.70 and below, it does this if
.I s
is sticky.)
.B svscan
needs two free descriptors for each pipe.

If
.I log-service
is specified and the named subdirectory exists,
.B svscan
starts the service found there and redirects its output through it.
This service is started before any other.

Every five seconds,
.B svscan
checks for subdirectories again. If it sees a new subdirectory, it starts a
new
.BR supervise (8)
process. If it sees an old subdirectory where a
.BR supervise (8)
process has exited, it restarts the
.BR supervise (8)
process. In the
.B log
case it reuses the same pipe so that no data is lost.

.B svscan
is designed to run forever. If it has trouble creating a pipe or running
.BR supervise (8),
it prints a message to stderr; it will try again five seconds later.

If
.B svscan
is given a command-line argument
.IR directory ,
it switches to that
.I directory
when it starts.

If
.B svscan
receives a TERM signal, it sends TERM signals to all non-log child
.BR supervise (8)
processes and the five second service scan is reduced to one second. New
services are ignored and if a running
.BR supervise (8)
stops, it is not restarted. As each existing
.BR supervise (8)
exits,
.B svscan
sends a TERM signal to the corresponding log
.BR supervise (8),
if one exists. After all child processes exit,
.B svscan
exits.

.SH SEE ALSO
envdir(8),
envini(8),
envuidgid(8),
fghack(8),  
multilog(8),
pgrphack(8),
readproctitle(8),
setlock(8),
setuidgid(8),
setuser(8),
softlimit(8),
supervise(8),
svc(8),
svok(8),
svscanboot(8),
svstat(8),
tai64n(8),
tai64nlocal(8)


CHANGES
Enhanced svscan to stop its services when it is stopped.
(Thanks to Rick Myers)

Added new program setuser to setuid, setenv, plus chdir before exec.

Added new program envini to read environment variables from an INI file.

Added support for adding a prefix in envdir.

Added "once" flag output to svstat.

Fixed supervise to run "start" on down services and "svc -o".

Fixed sleeper to do a single atomic write.

Added support for resetting limits in softlimit.

Fixed initial supervise "down" status.

Fix supervise removing "status" before locking the service.

Fix initialization of uptime on down service with log.

Changes in daemontools-encore-1.10
==================================

Added support for fnmatch patterns in multilog.

Increased the multilog file size limit to 2GB.
(Thanks to SATOH Fumiyasu)

Fixed missing separator between system error texts.

Major speed ups in multilog, tai64n, and tai64nlocal.

Added support for signal propagation and disabling setsid behavior in
supervise.  (Thanks to Alan Grow)

Changes in daemontools-encore-1.06
==================================

Added options to svstat to control checking the log script service.

Added -l option to svc to explicitly control the main service.

Fixed race issuing svc commands to log and then main service.

Changed the "safely written" option for multilog to "w".

Fixed daemontools leaking the status file descriptor to its children.

Fixed svc sending commands to both the log and main service.

Changes in daemontools-encore-1.05
==================================

Added options to svup to control checking the log script service.

Eliminated some compiler warnings.

Changes in daemontools-encore-1.04
==================================

Fixed bug in svscan preventing it from starting services with a "log"
script instead of directory.

Changes in daemontools-encore-1.03
==================================

Added -s option to setuidgid to set the supplemental GIDs.
(Thanks to SATOH Fumiyasu)

Changes in daemontools-encore-1.02
==================================

If a supervise is managing both a main and log service, issuing
"svc -dx" on the service will cause the log to stop after the main
service exits.

Changes between daemontools-0.76 and daemontools-encore
=======================================================

Added support to supervise/svc for sending SIGQUIT/USR1/USR2/WINCH.
(original patch by Wayne Marshall, based on patch by Scott Gifford)

Modified supervise to automatically put the forked service into its own
process group to facilitate sending signals to the group.

Added support to supervise/svc for sending signals to the group.

Added support to softlimit for reading limits from environment variables.

Modified supervise to run "start" once before executing "run" to allow
for single-shot initialization, and to run "stop" after "run" exits for
the last time.  The presence of "run" is also optional.

Modified supervise to stop trying to restart the service if it exits 100.
(semantics borrowed from qmail control files)

Added support to multilog for "accustamp" time stamps.
(derived from patch by Matthias Andree)

Added support to supervise for storing the status files in an alternate
directory specified by $SUPERVISEDIR.

Modified multilog to support specifying an alternate "safely written"
filename extension (useful for compressed logs).

Added extended state information to the supervise/status file and svstat.

Added support to supervise for a "notify" script, executed whenever a
subprocess is started or exits.  If present, it is passed the script
name, "start" or "exit", the PID, and the exit code.

Added support for piping to "log" within supervise.  This requires "log"
to be an executable file (instead of a directory as for svscan).

Modified supervise to overwrite status instead of creating a new file.
This allows the "supervise" subdirectory to be non-writeable.

Added a "svup" tool to determine if a service is running.

Modified svscan to optionally redirect all output through a service.
(adapted from http://www.gluelogic.com/code/daemontools/daemontools-0.76-gps-logging-0.01.patch)
Note that this feature works differently from the GlueLogic
modification.  See the svscan(8) man page for details.


svscan.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include "sig.h"
#include "direntry.h"
#include "strerr.h"
#include "error.h"
#include "open.h"
#include "wait.h"
#include "closeonexec.h"
#include "fd.h"
#include "env.h"
#include "str.h"
#include "byte.h"
#include "pathexec.h"

#define SERVICES 1000

#define INFO "svscan: info: "
#define WARNING "svscan: warning: "
#define FATAL "svscan: fatal: "

struct {
  unsigned long dev;
  unsigned long ino;
  int flagactive;
  int flaglog;
  int pid; /* 0 if not running */
  int pidlog; /* 0 if not running */
  int pi[2]; /* defined if flaglog */
} x[SERVICES];
int numx = 0;
int logx = -1;
int logpipe[2];
const char *logdir = 0;
int all_stopped, exit_asap=0;

char fnlog[260];

void catch_sig(int sig)
{
  exit_asap = 1;
}

void start(const char *fn)
{
  unsigned int fnlen;
  struct stat st;
  int child;
  int i;
  const char *args[3];
  int islog;

  islog = logdir && !str_diff(fn,logdir);
  if (fn[0] == '.' && !islog) return;

  if (stat(fn,&st) == -1) {
    strerr_warn4sys(WARNING,"unable to stat ",fn,"");
    return;
  }

  if ((st.st_mode & S_IFMT) != S_IFDIR) return;

  for (i = 0;i < numx;++i)
    if (x[i].ino == st.st_ino)
      if (x[i].dev == st.st_dev)
	break;

  if (i == numx) {
    if (numx >= SERVICES) {
      strerr_warn4(WARNING,"unable to start ",fn,": running too many services",0);
      return;
    }
    x[i].ino = st.st_ino;
    x[i].dev = st.st_dev;
    x[i].pid = 0;
    x[i].pidlog = 0;
    x[i].flaglog = 0;

    fnlen = str_len(fn);
    if (fnlen <= 255) {
      byte_copy(fnlog,fnlen,fn);
      byte_copy(fnlog + fnlen,5,"/log");
      if (stat(fnlog,&st) == 0)
	x[i].flaglog = S_ISDIR(st.st_mode);
      else
	if (errno != error_noent) {
          strerr_warn4sys(WARNING,"unable to stat ",fn,"/log");
          return;
	}
    }

    if (x[i].flaglog) {
      if (pipe(x[i].pi) == -1) {
        strerr_warn4sys(WARNING,"unable to create pipe for ",fn,"");
        return;
      }
      closeonexec(x[i].pi[0]);
      closeonexec(x[i].pi[1]);
    }
    ++numx;
  }

  x[i].flagactive = 1;

  if (!x[i].pid && !exit_asap)
    switch(child = fork()) {
      case -1:
        strerr_warn4sys(WARNING,"unable to fork for ",fn,"");
        return;
      case 0:
        if (x[i].flaglog)
	  if (fd_move(1,x[i].pi[1]) == -1)
            strerr_die3sys(111,WARNING,"unable to set up descriptors for ",fn);
	if (i == logx)
	  if (fd_move(0,logpipe[0]) == -1)
	    strerr_die3sys(111,WARNING,"unable to set up descriptors for ",fn);
        args[0] = "supervise";
        args[1] = fn;
        args[2] = 0;
	pathexec_run(*args,args,(const char*const*)environ);
        strerr_die3sys(111,WARNING,"unable to start supervise ",fn);
      default:
	x[i].pid = child;
        all_stopped = 0;
    }
  else
    if (x[i].pid) {
      all_stopped = 0;
      if (exit_asap)
        kill(x[i].pid,SIGTERM);
    }

  if (x[i].flaglog && !x[i].pidlog && !exit_asap)
    switch(child = fork()) {
      case -1:
        strerr_warn4sys(WARNING,"unable to fork for ",fn,"/log");
        return;
      case 0:
        if (fd_move(0,x[i].pi[0]) == -1)
          strerr_die4sys(111,WARNING,"unable to set up descriptors for ",fn,"/log");
	if (chdir(fn) == -1)
          strerr_die3sys(111,WARNING,"unable to switch to ",fn);
        args[0] = "supervise";
        args[1] = "log";
        args[2] = 0;
	pathexec_run(*args,args,(const char*const*)environ);
        strerr_die4sys(111,WARNING,"unable to start supervise ",fn,"/log");
      default:
	x[i].pidlog = child;
        all_stopped = 0;
    }
  else
    if (x[i].flaglog && x[i].pidlog) {
      all_stopped = 0;
      if (exit_asap && !x[i].pid)
        kill(x[i].pidlog,SIGTERM);
    }
}

void direrror(void)
{
  strerr_warn2sys(WARNING,"unable to read directory");
}

void doit(void)
{
  DIR *dir;
  direntry *d;
  int i;
  int r;
  int wstat;

  for (;;) {
    r = wait_nohang(&wstat);
    if (!r) break;
    if (r == -1) {
      if (errno == error_intr) continue; /* impossible */
      break;
    }

    for (i = 0;i < numx;++i) {
      if (x[i].pid == r) { x[i].pid = 0; break; }
      if (x[i].pidlog == r) { x[i].pidlog = 0; break; }
    }
  }

  for (i = 0;i < numx;++i)
    x[i].flagactive = 0;

  dir = opendir(".");
  if (!dir) {
    direrror();
    return;
  }
  for (;;) {
    errno = 0;
    d = readdir(dir);
    if (!d) break;
    start(d->d_name);
  }
  if (errno) {
    direrror();
    closedir(dir);
    return;
  }
  closedir(dir);

  i = 0;
  while (i < numx) {
    if (!x[i].flagactive && !x[i].pid && !x[i].pidlog) {
      if (x[i].flaglog) {
        close(x[i].pi[0]);
        close(x[i].pi[1]);
        x[i].flaglog = 0;
      }
      x[i] = x[--numx];
      continue;
    }
    ++i;
  }
}

static void start_log(void)
{
  struct stat st;

  /* Make sure the standard file descriptors are open before creating any pipes. */
  if (fstat(0,&st) != 0 && errno == EBADF)
    (void) open_read("/dev/null");
  if (fstat(1,&st) != 0 && errno == EBADF)
    (void) open_write("/dev/null");
  if (fstat(2,&st) != 0 && errno == EBADF)
    (void) open_write("/dev/null");

  if (logdir && stat(logdir,&st) == 0 && S_ISDIR(st.st_mode)) {
    logx = numx;
    if (pipe(logpipe) == -1)
      strerr_die3sys(111,FATAL,"unable to create pipe for ",logdir);
    closeonexec(logpipe[0]);
    closeonexec(logpipe[1]);
    start(logdir);
    if (numx > logx && x[logx].pid != 0) {
      (void) fd_copy(1,logpipe[1]);
      (void) fd_move(2,logpipe[1]);
      strerr_warn2(INFO,"starting",0);
    }
  }
}

int main(int argc,char **argv)
{
  if (argc >= 2)
    if (chdir(argv[1]) == -1)
      strerr_die3sys(111,FATAL,"unable to chdir to ",argv[1]);
  if (argc >= 3)
    logdir = argv[2];

  start_log();
  sig_catch(SIGTERM,catch_sig);

  for (;;) {
    all_stopped = 1;
    doit();
    if (all_stopped) break;
    sleep(exit_asap ? 1 : 5);
  }
  return 0;
}
